src/bio/bio_dev_monitor.c

```diff
@@ -0,0 +1,442 @@
+	uint32_t					 error_page_buf_sz;
+	uint32_t					 numd, numdl, numdu;
+	int						 rc;
+	int						 sc, sct;
+
+	/* Additional NVMe status information */
+	spdk_bdev_io_get_nvme_status(bdev_io, &sct, &sc);
+	if (sc)
+		D_ERROR("NVMe status code/type: %d/%d\n", sc, sct);
+
+	/* Free I/O request in the competion callback */
+	spdk_bdev_free_io(bdev_io);
+
+	health_desc = dev_health->bhm_desc;
+	channel = spdk_bdev_get_io_channel(health_desc);
+	if (!channel) {
+		D_ERROR("Failed to get bdev I/O channel for health desc\n");
+		return;
+	}
+
+	bdev = spdk_bdev_desc_get_bdev(health_desc);
+	cdata = dev_health->bhm_ctrlr_buf;
+	/* set the number of error log page entries supported by device */
+	dev_health->bhm_elpe = cdata->elpe;
+
+
+	/*TODO Add additional relevant controller data */
+	D_PRINT("==========================================================\n");
+	D_PRINT("SPDK Device Controller Data [%s]:\n",
+		bdev != NULL ? spdk_bdev_get_name(bdev) : "");
+	D_PRINT("==========================================================\n");
+	D_PRINT("Vendor ID: %04x\n", cdata->vid);
+	D_PRINT("Serial Number: ");
+	dprint_ascii_string(cdata->sn, sizeof(cdata->sn));
+	D_PRINT("\n");
+	D_PRINT("Model Number: ");
+	dprint_ascii_string(cdata->mn, sizeof(cdata->mn));
+	D_PRINT("\n");
+	D_PRINT("Firmware Version: ");
+	dprint_ascii_string(cdata->fr, sizeof(cdata->fr));
+	D_PRINT("\n");
+	D_PRINT("Error log page entries supported: %d\n", cdata->elpe + 1);
+
+	/* Prep NVMe command to get device error log pages */
+	error_page_sz = sizeof(struct spdk_nvme_error_information_entry);
+	numd = error_page_sz / sizeof(uint32_t) - 1u;
+	numdl = numd & 0xFFFFu;
+	numdu = (numd >> 16) & 0xFFFFu;
+	memset(&cmd, 0, sizeof(cmd));
+	cmd.opc = SPDK_NVME_OPC_GET_LOG_PAGE;
+	cmd.nsid = SPDK_NVME_GLOBAL_NS_TAG;
+	cmd.cdw10 = numdl << 16;
+	cmd.cdw10 |= SPDK_NVME_LOG_ERROR;
+	cmd.cdw11 = numdu;
+	error_page_buf_sz = error_page_sz * (dev_health->bhm_elpe + 1);
+
+	/*
+	 * Submit an NVMe Admin command to get device error log page
+	 * to the bdev.
+	 */
+	rc = spdk_bdev_nvme_admin_passthru(health_desc, channel, &cmd,
+					   dev_health->bhm_error_buf,
+					   error_page_buf_sz,
+					   get_spdk_err_log_page_completion,
+					   dev_health);
+	if (rc)
+		D_ERROR("NVMe admin passthru (error log), rc:%d\n", rc);
+}
+
+static void
+get_spdk_log_page_completion(struct spdk_bdev_io *bdev_io, bool success,
+			     void *cb_arg)
+{
+	struct bio_health_monitoring			*dev_health = cb_arg;
+	struct spdk_nvme_health_information_page	*health_page;
+	struct spdk_io_channel				*channel;
+	struct spdk_bdev				*bdev;
+	struct spdk_bdev_desc				*health_desc;
+	struct spdk_nvme_cmd				 cmd;
+	uint32_t					 ctrlr_page_sz;
+	int						 rc;
+	int						 sc, sct;
+
+	/* Additional NVMe status information */
+	spdk_bdev_io_get_nvme_status(bdev_io, &sct, &sc);
+	if (sc)
+		D_ERROR("NVMe status code/type: %d/%d\n", sc, sct);
+
+	/* Free I/O request in the competion callback */
+	spdk_bdev_free_io(bdev_io);
```

Comment:
Shouldn't the 'bdev_io' being freed only on error or on the last completion chain call?

Reply:
I checked spdk code, looks you are right, the spdk_bdev_free_io() should be called on each completion, then I'm wondering why it's only called on last completion in spdk/lib/bdev/nvme/bdev_nvme_rpc.c ? Probably a bug in SPDK?